XPCBackupService/Entities/BackupTreeNode.swift

```diff
@@ -89,48 +95,89 @@ class BackupTreeNode {
         return attribute[FileAttributeKey.modificationDate] as? Date
     }
 
-    private func nodeIsSynced(url: String, deviceId: Int) throws -> Bool {
-        let realm = try BackupRealm.shared.getRealm()
-        let syncedNode = realm.objects(SyncedNode.self).first { syncedNode in
-            url == syncedNode.url && deviceId == syncedNode.deviceId
+    private func nodeIsSynced(url: URL, deviceId: Int) throws -> ThreadSafeReference<SyncedNode>? {
+        let syncedNode = autoreleasepool {
+            let syncedNode = BackupRealm.shared.findSyncedNode(url: url, deviceId: deviceId)
+            
+            return syncedNode
         }
-
-        guard let syncedNodeDate = syncedNode?.updatedAt, let fileModificationDate = try self.getFileModificationDate() else {
-            return false
+        
+        guard let syncedNodeUnwrapped = syncedNode else {
+            return nil
+        }
+        
+        let syncedNodeDate = syncedNodeUnwrapped.updatedAt
+        
+        guard let fileModificationDate = try self.getFileModificationDate() else {
+            return nil
         }
 
         if (syncedNodeDate < fileModificationDate && self.type != .folder) {
             self.syncStatus = .NEEDS_UPDATE
             self.remoteId = syncedNode?.remoteId
             self.remoteUuid = syncedNode?.remoteUuid
             
-            return false
+            return nil
         }
+        
 
-        return syncedNode != nil
+        return ThreadSafeReference(to: syncedNodeUnwrapped)
```

Comment:
@patricioxavier8 `ThreadSafeReferences` allows us to pass Realm references between different threads, Realm does not allow us to perform thread switching with operations, an example:

If we create a `SyncedNode` in Thread #1, we can only operate with that reference in that thread, if we try to operate with that reference in another thread (let's say a delete operation), Realm will crash with a "realm accessed from incorrect thread" error. To safely share references between threads we use a `ThreadSafeReference`.

Documentation https://www.mongodb.com/docs/realm-sdks/swift/latest/Structs/ThreadSafeReference.html

Reply:
good to know, thanks for the explanation